database/src/main/scala/cromwell/database/SqlDatabase.scala

```diff
@@ -178,22 +178,12 @@ trait SqlDatabase extends AutoCloseable {
                                                   (implicit ec: ExecutionContext):
   Future[Traversable[(Execution, ExecutionInfo)]]
 
-  protected def addMetadataEvent(workflowUuid: String,
-                                 key: String,
-                                 value: String,
-                                 valueType: String,
-                                 timestamp: Timestamp)
-                                (implicit ec: ExecutionContext): Future[Unit]
-
-  protected def addMetadataEvent(workflowUuid: String,
-                                 key: String,
-                                 callFqn: String,
-                                 index: Option[Int],
-                                 attempt: Int,
-                                 value: String,
-                                 valueType: String,
-                                 timestamp: Timestamp)
-                                (implicit ec: ExecutionContext): Future[Unit]
+  /**
+    * Add metadata events to the database transactionally. normalized type structure is as follows:
+    * (WorkflowId, MetadataKey, Option[CallFqn, CallIndex, CallAttempt], MetadataValue, MetadataValueType, Timestamp)
+    */
+  protected def addMetadata(events: Iterable[(String, String, Option[(String, Option[Int], Int)], String, String, Timestamp)])
+                           (implicit ec: ExecutionContext): Future[Unit]
```

Comment:
Very-ToL:

a)

While I'm ok with all the discussions above, my preference here is that there be more than one overloaded `addMetadata()`, instead of the slick implementation having to peek so much at the `Some` or `None` passed in. One would only work on calls, the other on the workflow level. The names may need to be changed to `addWorkflowMetadata(Iterable[_], …)` and `addCallMetadata(Iterable[_], …)` due to the persistently annoying java erasure.

b)

I don't mind additional case classes in `cromwell.database.obj` for creating or returning data from the database. I do prefer the above mentioned overloads with basic scala types, but a case class isn't what I'm really worried about bleeding back into the slick code. Pre-`cromwell.database`, what I feel like I saw was: case classes with all sorts of conversion or business logic methods added, especially implicit converters in companion objects. Then all those methods would be invoked inside slick specific code, instead of basic CRUD, making it really hard to separate out what was actually slick FRM, and what was logic.

My latest 2¢